subsys/net/ip/tcp2.c

```diff
@@ -832,6 +954,89 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
 	return conn;
 }
 
+static void trim_unacked_data(struct tcp *conn)
+{
+	/* Make sure there is no empty buffers in the beginning
+	 * of the chain as we just pulled data from it (so there
+	 * should not be any)
+	 */
+	if (conn->unacked_data->buffer &&
+	    conn->unacked_data->buffer->len == 0) {
+		/* TODO: Fix net_pkt_pull() so that this check can be removed.
+		 */
+		struct net_buf *buf = conn->unacked_data->buffer;
+
+		while (buf) {
+			if (buf->len) {
+				break;
+			}
+
+			conn->unacked_data->buffer = buf->frags;
+			buf->frags = NULL;
+
+			net_buf_unref(buf);
+			if (!conn->unacked_data->buffer) {
+				break;
+			}
+
+			buf = conn->unacked_data->buffer->frags;
+		}
+	}
+}
+
+static u32_t calc_acked_data(struct tcp *conn, struct tcphdr *th)
+{
+	u32_t acked_len = th_ack(th) - conn->last_seq;
+
+	if (!conn->unacked_data || acked_len == 0) {
+		return 0U;
+	}
+
+	NET_DBG("Peer acked %u bytes", acked_len);
+
+	return acked_len;
+}
+
+static void update_acked_data(struct tcp *conn, u32_t acked_len)
```

Comment:
update_acked_data(struct tcp *conn, u32_t acked_len) -> tcp_send_data_remove(conn, len)?

I propose still rework it into the function that deals with the pull and trim and the rest that would deal with unacked_len and timer.